server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongRareTermsAggregator.java

```diff
@@ -0,0 +1,204 @@
+
+    static final BucketOrder ORDER = BucketOrder.compound(BucketOrder.count(true), BucketOrder.key(true)); // sort by count ascending
+
+    //TODO review question: is LongLong map ok?
+    protected LongLongHashMap map;
+    protected LongHash bucketOrds;
+
+    private LeafBucketCollector subCollectors;
+
+    LongRareTermsAggregator(String name, AggregatorFactories factories, ValuesSource.Numeric valuesSource, DocValueFormat format,
+                                   SearchContext aggregationContext, Aggregator parent, IncludeExclude.LongFilter longFilter,
+                                   int maxDocCount, List<PipelineAggregator> pipelineAggregators,
+                                   Map<String, Object> metaData) throws IOException {
+        super(name, factories, aggregationContext, parent, pipelineAggregators, metaData, maxDocCount, format, valuesSource, longFilter);
+        this.map = new LongLongHashMap();
+        this.bucketOrds = new LongHash(1, aggregationContext.bigArrays());
+    }
+
+    protected SortedNumericDocValues getValues(ValuesSource.Numeric valuesSource, LeafReaderContext ctx) throws IOException {
+        return valuesSource.longValues(ctx);
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+                                                final LeafBucketCollector sub) throws IOException {
+        final SortedNumericDocValues values = getValues(valuesSource, ctx);
+        if (subCollectors == null) {
+            subCollectors = sub;
+        }
+        return new LeafBucketCollectorBase(sub, values) {
+            private long numDeleted = 0;
+
+            @Override
+            public void collect(int docId, long owningBucketOrdinal) throws IOException {
+                if (values.advanceExact(docId)) {
+                    final int valuesCount = values.docValueCount();
+
+                    long previous = Long.MAX_VALUE;
+                    for (int i = 0; i < valuesCount; ++i) {
+                        final long val = values.nextValue();
+                        if (previous != val || i == 0) {
+                            if ((includeExclude == null) || (includeExclude.accept(val))) {
+                                if (bloom.mightContain(val) == false) {
+                                    long termCount = map.get(val);
+                                    if (termCount == 0) {
+                                        // Brand new term, save into map
+                                        map.put(val, 1L);
+                                        long bucketOrdinal = bucketOrds.add(val);
+                                        if (bucketOrdinal < 0) { // already seen
+                                            bucketOrdinal = - 1 - bucketOrdinal;
+                                            collectExistingBucket(subCollectors, docId, bucketOrdinal);
```

Comment:
Would getting here not indicate a bug since we have determined the `termCount` to be `0` already so we should not have seen the bucket ordinal before?